File: pylib/anki/pgrep/card_sets.py
# Copyright: Ankitects Pty Ltd and contributors
# License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html

"""Card Sets read model for the Library "wheel" browser (L4).

Groups the learner's flash cards into one **set per blueprint category**, in
blueprint order, for the Card Sets browser (``design/claude-design/
design_handoff_card_sets``). A "set" is a category (Classical Mechanics, E&M,
...); its cards are the ``Basic`` notes tagged ``topic::<category>[::<sub>]``
across the seeded ``PGRE::Sample`` deck and the authored / AI ``PGRE::Generated``
deck.

Honesty rule (``api-contract.md`` §0): every count and preview is real. A
set's card count is the true number of notes, and the deck-face preview
(``cards[0].front``) is the first card's actual front. Categories with no cards
are omitted, so the wheel never shows an invented set. No AI, and no scheduling
state is read or written.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any

from anki.pgrep.blueprint import CATEGORY_SLUGS
from anki.pgrep.tags import TOPIC_PREFIX, category_for

if TYPE_CHECKING:
    from anki.collection import Collection

# The decks a card set draws from: the seeded sample deck and the authored /
# AI-generated deck. Both hold ``Basic`` topic-tagged notes.
SET_DECKS: tuple[str, ...] = ("PGRE::Sample", "PGRE::Generated")

# Category slug -> display name. Mirrors the Home manifold labels
# (``anki.pgrep.manifold``) so a category reads the same on every surface. This
# is duplicated rather than imported: the manifold's layout table is presentation
# coupled to the 3D map, not a shared naming source (see ``blueprint.py`` on
# intentional per-boundary duplication).
CATEGORY_NAMES: dict[str, str] = {
    "mechanics": "Classical Mechanics",
    "electromagnetism": "Electromagnetism",
    "quantum": "Quantum Mechanics",
    "thermodynamics": "Thermo & Stat Mech",
    "atomic": "Atomic Physics",
    "optics_waves": "Optics & Waves",
    "special_relativity": "Special Relativity",
    "lab": "Laboratory Methods",
    "specialized": "Specialized Topics",
}


def _display_name(category: str) -> str:
    """Human name for a category slug (Title-Cased fallback for the unexpected)."""
    return CATEGORY_NAMES.get(category, category.replace("_", " ").title())


def list_card_sets(col: Collection) -> list[dict[str, Any]]:
    """Return the learner's card sets, one per blueprint category with cards.

    Each entry is ``{"category", "name", "cards": [{"note_id", "front",
    "back"}, ...]}``. Categories come back in blueprint order
    (:data:`anki.pgrep.blueprint.CATEGORY_SLUGS`); a category with no cards is
    omitted. Cards keep note-id (insertion) order, so the deck-face preview
    (``cards[0].front``) is stable. JSON-serializable; no AI, no scheduler.
    """
    decks = " OR ".join(f'deck:"{deck}"' for deck in SET_DECKS)
    query = f"note:Basic ({decks}) tag:{TOPIC_PREFIX}*"

    by_category: dict[str, list[dict[str, Any]]] = {}
    for nid in sorted(col.find_notes(query)):
        note = col.get_note(nid)
        # Guard the field read: a stray non-Basic match would lack Front/Back.
        if "Front" not in note or "Back" not in note:
            continue
        category = category_for(note.tags)
        by_category.setdefault(category, []).append(
            {"note_id": int(nid), "front": note["Front"], "back": note["Back"]}
        )

    # Blueprint order first; any unrecognized category (should not occur for
    # topic-tagged content) is appended alphabetically so nothing is dropped.
    ordered = [c for c in CATEGORY_SLUGS if c in by_category]
    ordered += sorted(c for c in by_category if c not in CATEGORY_SLUGS)

    return [
        {
            "category": category,
            "name": _display_name(category),
            "cards": by_category[category],
        }
        for category in ordered
    ]


def add_card(col: Collection, category: str, front: str, back: str) -> dict[str, Any]:
    """Author one card into a category's set, as-is (no AI). Returns the new id.

    This is the wheel's "Add a card": the learner's own front/back go straight
    into the ``PGRE::Generated`` deck tagged for ``category``, via
    :func:`anki.pgrep.generation.author_seed` (the generation-effect authoring
    path, which works with AI off and imports no AI modules). Distinct from
    calibration, which is the guided walkthrough. Returns ``{"note_id",
    "category"}``.
    """
    from anki.pgrep import generation

    result = generation.author_seed(col, front.strip(), back.strip(), category)
    return {"note_id": int(result["note_id"]), "category": category}

File: mobile/ios/PgrepStudy/CardSets.swift
// Copyright: Ankitects Pty Ltd and contributors
// License: GNU AGPL, version 3 or later; http://www.gnu.org/licenses/agpl.html
//
// The Library / Card Sets read model + "Add a card" write path for the iOS
// companion. A faithful port of pylib/anki/pgrep/card_sets.py (list_card_sets +
// add_card, which delegates to generation.author_seed):
//
//   - list_card_sets: group the learner's Basic topic-tagged cards in the seeded
//     PGRE::Sample and authored PGRE::Generated decks into one set per blueprint
//     category, in blueprint order (unrecognized categories appended
//     alphabetically), dropping categories with no cards. Real counts and a real
//     deck-face preview (cards[0].front); no AI, no scheduler.
//   - add_card -> author_seed: the learner's own front/back go straight into the
//     PGRE::Generated deck as a new Basic note tagged [pgrep::seed-authored,
//     topic::<category>], so a card added on the phone is the same kind of note
//     as a desktop-added one and merges on sync, landing in the right set and
//     (once studied) Memory / Coverage.
//
// The grouping / ordering / naming is pure and unit-testable (see CardSets.group
// and CardSetsTests); the engine reads and writes via the note RPCs in the
// AnkiBackend extension below, mirroring how AttemptWriter.swift adds notes.
//
// The blueprint order and the category display names are duplicated here on
// purpose, the same cross-language boundary duplication the L1 contract mandates
// (docs_pgrep/contracts/L1-coordination-schema.md §1); the order reuses the
// Blueprint table already ported in MemoryScore.swift.

import Foundation

/// One browsable card in a set. Mirrors list_card_sets' `{note_id, front, back}`.
struct CardSetCard: Sendable, Equatable, Identifiable {
    var noteId: Int64
    var front: String
    var back: String

    var id: Int64 { noteId }
}

/// One category's set of cards. Mirrors a list_card_sets entry
/// (`{category, name, cards}`).
struct CardSet: Sendable, Equatable, Identifiable {
    var category: String
    var name: String
    var cards: [CardSetCard]

    var id: String { category }
}

enum CardSets {
    /// The decks a card set draws from: the seeded sample deck and the authored /
    /// AI-generated deck. Mirrors card_sets.SET_DECKS.
    static let setDecks = ["PGRE::Sample", "PGRE::Generated"]

    /// The deck a learner-authored card lands in. Mirrors
    /// generation.GENERATED_DECK_NAME.
    static let generatedDeckName = "PGRE::Generated"

    /// The seed-authored marker tag on every authored card. Mirrors
    /// generation.SEED_TAG (distinct from the seeder's pgrep::seeded).
    static let seedTag = "pgrep::seed-authored"

    /// The Basic notetype a card set's notes (and authored cards) use. Mirrors
    /// generation._basic / seed's col.models.by_name("Basic").
    static let basicNotetypeName = "Basic"

    /// Category slug -> display name. Mirrors card_sets.CATEGORY_NAMES so a
    /// category reads the same on every surface. Duplicated rather than derived
    /// from the manifold labels on purpose (that table is presentation-coupled to
    /// the 3D map, per blueprint.py on intentional per-boundary duplication).
    static let categoryNames: [String: String] = [
        "mechanics": "Classical Mechanics",
        "electromagnetism": "Electromagnetism",
        "quantum": "Quantum Mechanics",
        "thermodynamics": "Thermo & Stat Mech",
        "atomic": "Atomic Physics",
        "optics_waves": "Optics & Waves",
        "special_relativity": "Special Relativity",
        "lab": "Laboratory Methods",
        "specialized": "Specialized Topics",
    ]

    /// Human name for a category slug (Title-Cased fallback for the unexpected).
    /// Mirrors card_sets._display_name.
    static func displayName(for category: String) -> String {
        if let name = categoryNames[category] { return name }
        return category
            .split(separator: "_", omittingEmptySubsequences: true)
            .map { $0.prefix(1).uppercased() + $0.dropFirst().lowercased() }
            .joined(separator: " ")
    }

    /// The topic tag for a category slug. Mirrors generation._topic_tag: a bare
    /// slug becomes "topic::<slug>"; an already-prefixed tag is kept verbatim.
    static func topicTag(for category: String) -> String {
        category.lowercased().hasPrefix(Topic.prefix) ? category : "\(Topic.prefix)\(category)"
    }

    /// The Anki search that finds a set's candidate notes. Mirrors
    /// list_card_sets' query exactly: Basic notes in the set decks that carry a
    /// topic tag.
    static var searchQuery: String {
        let decks = setDecks.map { "deck:\"\($0)\"" }.joined(separator: " OR ")
        return "note:\(basicNotetypeName) (\(decks)) tag:\(Topic.prefix)*"
    }

    /// A candidate note read from the engine, before grouping.
    struct NoteRow: Sendable, Equatable {
        var noteId: Int64
        var tags: [String]
        var front: String
        var back: String
    }

    /// Group candidate notes into one set per category, in blueprint order (any
    /// unrecognized category appended alphabetically so nothing is dropped),
    /// omitting categories with no cards. A line-for-line port of
    /// list_card_sets' grouping. Rows are sorted by note id first, so each set's
    /// `cards[0]` is the stable, insertion-order deck-face preview.
    static func group(rows: [NoteRow]) -> [CardSet] {
        var byCategory: [String: [CardSetCard]] = [:]
        for row in rows.sorted(by: { $0.noteId < $1.noteId }) {
            let category = Topic.category(forTags: row.tags)
            byCategory[category, default: []].append(
                CardSetCard(noteId: row.noteId, front: row.front, back: row.back)
            )
        }

        var ordered = Blueprint.slugs.filter { byCategory[$0] != nil }
        ordered += byCategory.keys.filter { !Blueprint.slugs.contains($0) }.sorted()

        return ordered.map { category in
            CardSet(
                category: category,
                name: displayName(for: category),
                cards: byCategory[category] ?? []
            )
        }
    }
}

extension AnkiBackend {
    /// Read the learner's card sets, a port of card_sets.list_card_sets over the
    /// note RPCs: find the Basic topic-tagged notes in the set decks, read each
    /// note's tags/front/back, and group them into one set per category in
    /// blueprint order. Note ids are sorted ascending so `cards[0]` is the stable
    /// deck-face preview, matching desktop's `sorted(find_notes(...))`. A stray
    /// match without both fields is skipped (the desktop Front/Back guard).
    func loadCardSets() throws -> [CardSet] {
        let noteIds = try searchNotes(matching: CardSets.searchQuery).sorted()
        var rows: [CardSets.NoteRow] = []
        rows.reserveCapacity(noteIds.count)
        for nid in noteIds {
            let note = try getNote(noteId: nid)
            // Guard the field read: a stray non-Basic match would lack Front/Back.
            guard note.fields.count >= 2 else { continue }
            rows.append(CardSets.NoteRow(
                noteId: note.id,
                tags: note.tags,
                front: note.fields[0],
                back: note.fields[1]
            ))
        }
        return CardSets.group(rows: rows)
    }

    /// Author one card into a category's set, as-is (no AI). A port of
    /// card_sets.add_card -> generation.author_seed: a new Basic note whose
    /// trimmed Front/Back go straight into the PGRE::Generated deck, tagged
    /// [pgrep::seed-authored, topic::<category>]. Returns the new note id. The
    /// notetype and deck are resolved (and bootstrapped only on a never-synced
    /// phone that lacks them) exactly like the desktop seed path, so an
    /// iOS-authored card is the same kind of note as a desktop one.
    @discardableResult
    func addCard(category: String, front: String, back: String) throws -> Int64 {
        let notetypeId = try ensureBasicNotetype()
        let deckId = try ensureGeneratedDeck()

        var note = Anki_Notes_Note()
        note.notetypeID = notetypeId
        note.fields = [
            front.trimmingCharacters(in: .whitespacesAndNewlines),
            back.trimmingCharacters(in: .whitespacesAndNewlines),
        ]
        note.tags = [CardSets.seedTag, CardSets.topicTag(for: category)]

        let response = try addNote(note, deckId: deckId)
        return response.noteID
    }

    /// The `Basic` notetype id, preferring the collection's existing (synced) one
    /// and only bootstrapping a schema-standard Basic on a never-synced phone
    /// that somehow lacks it. Mirrors AttemptWriter.ensureAttemptNotetype's
    /// find-then-create pattern; the shipped path always finds the synced Basic
    /// (the bundled sample collection ships with it, like desktop).
    func ensureBasicNotetype() throws -> Int64 {
        let existing = try notetypeId(forName: CardSets.basicNotetypeName)
        if existing != 0 { return existing }

        var notetype = Anki_Notetypes_Notetype()
        notetype.name = CardSets.basicNotetypeName
        var config = Anki_Notetypes_Notetype.Config()
        config.kind = .normal
        config.sortFieldIdx = 0
        notetype.config = config
        notetype.fields = ["Front", "Back"].map { name in
            var field = Anki_Notetypes_Notetype.Field()
            field.name = name
            return field
        }
        var template = Anki_Notetypes_Notetype.Template()
        template.name = "Card 1"
        template.config.qFormat = "{{Front}}"
        template.config.aFormat = "{{FrontSide}}\n\n<hr id=answer>\n\n{{Back}}"
        notetype.templates = [template]
        return try addNotetype(notetype)
    }

    /// The `PGRE::Generated` deck id, creating it (with its `PGRE` parent) when
    /// missing. Mirrors generation's `col.decks.id(GENERATED_DECK_NAME)`, which
    /// creates the deck on first author.
    func ensureGeneratedDeck() throws -> Int64 {
        let existing = try deckId(forName: CardSets.generatedDeckName)
        if existing != 0 { return existing }
        var deck = try newDeck()
        deck.name = CardSets.generatedDeckName
        return try addDeck(deck)
    }
}
